### PR TITLE
[WIP] Dapr Workflow initial methods for dotnet-sdk

### PR DESCRIPTION
**Pull request by amulyavarote**
# Description

This is a work in progress PR for dapr workflow initial methods for dotnet-sdk.
This needs e2e tests still.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #979 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

**Review by cgillum — COMMENTED** [review]
Based on all the other C# files I see in this repo, we'll need to add the common license header to all the new files we're adding in this PR.

File: src/Dapr.Workflow/ActivityContext.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DurableTask;
+
+namespace Microsoft.Extensions.DependencyInjection
```
Comment (cgillum) [inline src/Dapr.Workflow/ActivityContext.cs:6]: Change this to `Dapr.Workflow`

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+using Dapr.Workflow;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    /// <summary>
+    /// Defines runtime options for workflows.
+    /// </summary>
+    public sealed class WorkflowRuntimeOptions
+    {
+        // Dictionary<string, Action<IDurableTaskRegistry>> factories = new();
```
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowRuntimeOptions.cs:15]: Let's remove this commented-out line

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+using Microsoft.DurableTask;
+using Dapr.Workflow;
+
+namespace Microsoft.Extensions.DependencyInjection
```
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowRuntimeOptions.cs:8]: This namespace needs to be `Dapr.Workflow`.

File: src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.DurableTask.Grpc;
+using Microsoft.Extensions.DependencyInjection;
+using Dapr.Workflow;
+
+namespace Microsoft.Extensions.DependencyInjection
```
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowServiceCollectionExtensions.cs:8]: This namespace needs to be `Dapr.Workflow`.

File: src/Dapr.Workflow/WorkflowContext.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        /// Method to get the workflow name.
+        /// </summary>
+        public TaskName Name => this.innerContext.Name;
+        /// <summary>
+        /// Method to get the workflow id.
+        /// </summary>
+        public string InstanceId => this.innerContext.InstanceId;
+
+        /// <summary>
+        /// Method to get the current UTC Date time.
```
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowContext.cs:43]:
```suggestion
        /// Gets the current workflow time in UTC.
```

File: src/Dapr.Workflow/WorkflowContext.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        /// </summary>
+        public DateTime CurrentUtcDateTime => this.innerContext.CurrentUtcDateTime;
+
+        /// <summary>
+        /// Method to set the custom status to the workflow.
```
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowContext.cs:48]:
```suggestion
        /// Assigns a custom status value to the current workflow.
```

File: src/Dapr.Workflow/WorkflowContext.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        /// </summary>
+        public void SetCustomStatus(object? customStatus) => this.innerContext.SetCustomStatus(customStatus);
+
+        /// <summary>
+        /// Method to create a timer for the workflow.
```
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowContext.cs:53]:
```suggestion
        /// Creates a durable timer that expires after the specified delay.
```

File: src/Dapr.Workflow/WorkflowContext.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        /// </summary>
+        public Task CreateTimer(TimeSpan delay, CancellationToken cancellationToken = default)
+        {
+            return this.innerContext.CreateTimer(delay, cancellationToken);
+        }
+
+        /// <summary>
+        /// Method to wait for the external event.
```
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowContext.cs:61]:
```suggestion
        /// Waits for an event to be raised with name <paramref name="eventName"/> and returns the event data.
```

File: src/Dapr.Workflow/WorkflowContext.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        /// </summary>
+        public Task<T> WaitForExternalEventAsync<T>(string eventName, TimeSpan timeout)
+        {
+            return this.innerContext.WaitForExternalEvent<T>(eventName, timeout);
+        }
+
+        /// <summary>
+        /// Method to call the activity.
```
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowContext.cs:69]:
```suggestion
        /// Asynchronously invokes an activity by name and with the specified input value.
```

File: src/Dapr.Workflow/WorkflowMetadata.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ------------------------------------------------------------------------
+
+using Microsoft.DurableTask;
+
+namespace Dapr.Workflow
+{
+    /// <summary>
+    /// Defines properties and methods for workflow metadata.
```
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowMetadata.cs:19]:
```suggestion
    /// Represents a snapshot of a workflow instance's current state, including metadata.
```

File: src/Dapr.Workflow/WorkflowMetadata.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    /// </summary>
+    public class WorkflowMetadata
+    {
+        internal WorkflowMetadata(OrchestrationMetadata? metadata)
+        {
+            this.Details = metadata;
+        }
+        /// <summary>
+        /// Method to check if workflow exists.
```
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowMetadata.cs:28]:
```suggestion
        /// Gets a value indicating whether the requested workflow instance exists.
```

**Review by cgillum — COMMENTED** [review]
Adding more feedback, focusing mostly on documentation descriptions.

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    
+    /// <summary>
+    /// Defines runtime options for workflows.
+    /// </summary>
+    public sealed class WorkflowRuntimeOptions
+    {
+        /// <summary>
+        /// Dictionary to and name and registery of a workflow.
+        /// </summary>
+        public Dictionary<string, Action<IDurableTaskRegistry>> factories => new();
```
[the remark this answers is not in the document]
Comment (amulyavarote) [reply src/Dapr.Workflow/WorkflowRuntimeOptions.cs]: I tried to use readonly, but I got the below error. So kept it public.
`The modifier 'readonly' is not valid for this item [/Users/amulyavarote/Documents/Project/dotnet-sdk/src/Dapr.Workflow/Dapr.Workflow.csproj]`

[the remark this answers is not in the document]
Comment (cgillum) [reply src/Dapr.Workflow/WorkflowRuntimeOptions.cs]: This is the [exact syntax](https://github.com/cgillum/durabletask-dapr/blob/99aa3dbd87c57e9939125eda352742083f3dd336/src/Dapr.Workflow/WorkflowRuntimeOptions.cs#L10) I used in the prototype. Please double-check that you don't have any typos. This should absolutely *not* be public.

File: src/Dapr.Workflow/WorkflowClient.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        public async Task<WorkflowMetadata> GetWorkflowMetadataAsync(string instanceId, bool getInputsAndOutputs = false)
+        {
+            OrchestrationMetadata? metadata = await this.innerClient.GetInstanceMetadataAsync(
+                instanceId,
+                getInputsAndOutputs);
+            return new WorkflowMetadata(metadata);
+        }
+
+        /// <summary>
+        /// /// Disposes any unmanaged resources associated with this client.
```
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowClient.cs:64]: Need to remove the extra `///` here.

File: src/Dapr.Workflow/WorkflowRuntimeOptions.cs [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+    using Microsoft.DurableTask;
+    using Dapr.Workflow;
+    
+    /// <summary>
+    /// Defines runtime options for workflows.
+    /// </summary>
+    public sealed class WorkflowRuntimeOptions
+    {
+        /// <summary>
+        /// Dictionary to and name and registery of a workflow.
```
Comment (cgillum) [inline src/Dapr.Workflow/WorkflowRuntimeOptions.cs:29]: Looks like there is a typo in this summary text?

**Review by cgillum — COMMENTED** [review]
A couple tiny comment fixes but otherwise I think this looks good. I will try to make some time to run it locally to confirm that it's working.

**Comment by cgillum** [comment]
@amulyavarote I noticed that your new csproj files are not referenced in the `all.sln` file. Can you add these projects to the solution so that maintainers can easily find them?

**Comment by amulyavarote** [comment]
> 

Added!

**Comment by cgillum** [comment]
Thanks! It looks like that uncovered a build error which we would have otherwise missed. We may need to change the version of Microsoft.DurableTask.Client used in this PR. That, or we'll need to publish a 0.5.0 version to nuget.org.

Also, as far as .sln organization, I suggest you make it look like this so that it's consistent with the rest of the solution.
image

**Comment by cgillum** [comment]
Last thing I'll mention is that the workflow in the example project runs successfully for me. I'm not sure why it was failing for you.
```
INFO[1473] work item stream established by user-agent: [grpc-csharp/2.39.1 (.NET 6.0.11; CLR 6.0.11; netstandard2.0; x64) grpc-c/18.0.0 (windows; chttp2)]  app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.wfengine type=log ver=edge
INFO[1524] b0cfe42e: starting new 'HelloWorld' instance.  app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.wfengine type=log ver=edge
INFO[1524] Found reminder with key: dapr.internal.wfengine.workflow||b0cfe42e||start-557f1382. Deleting reminder  app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.actor type=log ver=edge
INFO[1524] Found reminder with key: dapr.internal.wfengine.activity||b0cfe42e#0||Execute. Deleting reminder  app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.actor type=log ver=edge
INFO[1524] b0cfe42e: 'HelloWorld' completed with a COMPLETED status.  app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.wfengine type=log ver=edge
INFO[1524] Found reminder with key: dapr.internal.wfengine.workflow||b0cfe42e||new-event-d855d47a. Deleting reminder  app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.actor type=log ver=edge
INFO[9959] work item stream closed                       app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.wfengine type=log ver=edge
```

**Comment by amulyavarote** [comment]
> Last thing I'll mention is that the workflow in the example project runs successfully for me. I'm not sure why it was failing for you.
> 
> ```
> INFO[1473] work item stream established by user-agent: [grpc-csharp/2.39.1 (.NET 6.0.11; CLR 6.0.11; netstandard2.0; x64) grpc-c/18.0.0 (windows; chttp2)]  app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.wfengine type=log ver=edge
> INFO[1524] b0cfe42e: starting new 'HelloWorld' instance.  app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.wfengine type=log ver=edge
> INFO[1524] Found reminder with key: dapr.internal.wfengine.workflow||b0cfe42e||start-557f1382. Deleting reminder  app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.actor type=log ver=edge
> INFO[1524] Found reminder with key: dapr.internal.wfengine.activity||b0cfe42e#0||Execute. Deleting reminder  app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.actor type=log ver=edge
> INFO[1524] b0cfe42e: 'HelloWorld' completed with a COMPLETED status.  app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.wfengine type=log ver=edge
> INFO[1524] Found reminder with key: dapr.internal.wfengine.workflow||b0cfe42e||new-event-d855d47a. Deleting reminder  app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.actor type=log ver=edge
> INFO[9959] work item stream closed                       app_id=foo instance=CGILLUM-DEV2 scope=dapr.runtime.wfengine type=log ver=edge
> ```

It works for me as well now.

**Review by cgillum — APPROVED** [review]
With the latest changes, I think we're good now. I see there were some E2E test failures, but I'm guessing those might just be flaky tests since your changes wouldn't impact them. Maybe @artursouza or another maintainer can re-run the failed job to see if it passes on retry?

**Comment by nyemade-uversky** [comment]
@halspang and @artursouza Can we get this merged? I'm curious what the path forward is on this

**Comment by shubham1172** [comment]
Can this PR be closed?

**Comment by cgillum** [comment]
Yes, this can be closed.

**Comment by yash-nisar** [comment]
@halspang Can we close this PR ?

**Comment by halspang** [comment]
Closing in favor of https://github.com/dapr/dotnet-sdk/pull/981

**Pull request closed without merging**